src/contract.rs

```diff
@@ -0,0 +1,144 @@
+use crate::{Transaction, ValidationError};
+
+#[cfg(feature = "std")]
+use crate::StorageSlot;
+
+use fuel_crypto::Hasher;
+use fuel_types::{Bytes32, ContractId, Salt};
+
+#[cfg(feature = "std")]
+use fuel_merkle::{binary::MerkleTree, common::StorageMap};
+
+#[cfg(feature = "std")]
+use fuel_types::Bytes64;
+
+use alloc::vec::Vec;
+
+#[cfg(feature = "std")]
+use core::cmp;
+
+#[derive(Debug, Default, Clone, PartialEq, Eq, Hash)]
+#[cfg_attr(feature = "serde", derive(serde::Serialize, serde::Deserialize))]
+/// Deployable representation of a contract code.
+pub struct Contract(Vec<u8>);
+
+impl Contract {
+    #[cfg(feature = "std")]
+    /// Calculate a binary merkle root with in-memory storage
+    fn ephemeral_merkle_root<L, I>(mut leaves: I) -> Bytes32
+    where
+        L: AsRef<[u8]>,
+        I: Iterator<Item = L>,
+    {
+        let mut storage = StorageMap::new();
+        let mut tree = MerkleTree::new(&mut storage);
+
+        // TODO fuel-merkle should have infallible in-memory struct
+        leaves
+            .try_for_each(|l| tree.push(l.as_ref()))
+            .and_then(|_| tree.root())
+            .expect("In-memory impl should be infallible")
+            .into()
+    }
+
+    #[cfg(feature = "std")]
+    /// Calculate the code root from a contract.
+    ///
+    /// <https://github.com/FuelLabs/fuel-specs/blob/master/specs/protocol/identifiers.md#contract-id>
+    pub fn root(&self) -> Bytes32 {
+        let root = self.0.chunks(8).map(|c| {
+            let mut bytes = [0u8; 8];
+
+            let l = cmp::min(c.len(), 8);
+            (&mut bytes[..l]).copy_from_slice(c);
+
+            bytes
+        });
+
+        Self::ephemeral_merkle_root(root)
+    }
+
+    #[cfg(feature = "std")]
+    /// Calculate the root of the initial storage slots for this contract
+    /// TODO: Use a sparse merkle tree once the implementation is available
```

Comment:
Can you file an issue for this?

Reply:
Is the comment correct in saying it should be SMT? Or is BMT fine?

Reply:
Storage root should be an SMT